lib/core/typeinfo.nim

```diff
@@ -341,6 +341,9 @@ proc getPointer*(x: Any): pointer =
   assert x.rawType.kind in pointerLike
   result = cast[ppointer](x.value)[]
 
+  if result != nil and x.rawType.kind != tyPointer:
+    result = cast[pointer](x.value)
```

Comment:
Still makes no sense to me. Maybe it's a marshal.nim bug?

Reply:
None of this code was written for clarity.  

I've just spent some time on it, and I now understand why there was an extra dereference happening.

The DEREFERENCED value (the actual memory location of the object) is returned to marshal via `getPointer()`, and this is what is cached *not* the pointer that the allocating `Any` set.

So the real issue is that, since `getPointer()` returns the actual reference to the memory the user is given, but `setPointer()` expects something different depending on how the any object has been marked.
If it's been marked as a pointer, it expects the actual reference to memory we need. Otherwise, it expects a POINTER to the reference. This is non-intuitive, and not clearly documented in `setPointer()`, but I am sure it's related to the underlying storage model for refs, which I did not dig into (any pointers to docs would be appreciated; I'd prefer not to dig through code).

When the second Any comes along and Marshal recognizes it's using the same object, it's passing to 'setPointer' the actual memory location, NOT a pointer to that pointer.

So the fix was doing the right thing for this one case, but yes, the real problem is that Marshal should have been storing the cached memory address in a local variable, and pass that, instead of passing the raw memory location.

So I've gone ahead and fixed it properly, and pushed that.

Again the asymmetry between `setPointer()` and  `getPointer()` clearly caused confusion and was unexpected. Some better documentation in typeinfo would be welcome-- it'd be nice to understand why it's storing a pointer to the object's address, instead of the object's address.

Reply:
A `ref` is implemented as a single `ptr` in Nim. The asymmetry in `setPointer` likely arises from the fact that `pointer` in Nim points to an object of unknown size.